GeneratorUI.py

```diff
@@ -0,0 +1,390 @@
+        root.withdraw()
+        return tkinter.filedialog.askopenfilename(
+            title=title,
+            filetypes=((t[0], " ".join(t[1])) for t in filetypes),
+            initialfile=suggest or None,
+        )
+
+
+class GeneratorApp(ThemedApp):
+    player_table: MDGridLayout
+    ui_log: UILog
+
+    players: list = []
+    dropped_text: list[str] = []
+    dropped_files: list[str] = []
+
+    def __init__(self):
+        super(GeneratorApp, self).__init__()
+        self.yamls = []
+
+    def build(self):
+        self.title = "Archipelago Generator"
+        Window.bind(on_drop_file=self._on_file_drop)
+        Window.bind(on_drop_text=self._on_text_drop)
+
+        Window.bind(on_drop_begin=self._on_drop_begin)
+        Window.bind(on_drop_end=self._on_drop_end)
+
+        ui = Builder.load_string(kv)
+        self.player_table = ui.ids.player_table
+        self.ui_log = ui.ids.ui_log
+
+        # for header in ["No.", "Name", "Game", "Remove"]:
+        self.player_table.add_widget(Label(text="No.", text_size=(None, 30), size_hint_x=2))
+        self.player_table.add_widget(Label(text="Name", text_size=(None, 30), size_hint_x=8))
+        self.player_table.add_widget(Label(text="Game", text_size=(None, 30), size_hint_x=8))
+        self.player_table.add_widget(Label(text="", text_size=(None, 30), size_hint_x=2))
+
+        return ui
+
+    def _on_drop_begin(self, *_) -> None:
+        self.dropped_text = []
+        self.dropped_files = []
+
+    def _on_file_drop(self, window, file_path: bytes, *_) -> None:
+        self.dropped_files.append(file_path.decode("utf-8"))
+
+    def _on_text_drop(self, window, text: bytes, *_) -> None:
+        self.dropped_text.append(text.decode("utf-8"))
+
+    def _on_drop_end(self, *_):
+        for file_path in self.dropped_files:
+            self._add_yaml_file(file_path)
+
+        for document in yaml.load_all(
+            "\n".join(self.dropped_text),
+            Loader=yaml.Loader,
+        ):
+            self._add_yaml(document)
+
+    def on_ref_press(self, _, ref: str) -> None:
+        match = re.match("file\\|(.*)", ref)
+        if match:
+            file_path = match.group(1)
+            asyncio.create_task(show_in_file_explorer(file_path))
+
+    def _add_yaml_file(self, file_path: str) -> None:
+        with open(file_path, "r", encoding="utf-8") as f:
+            document = yaml.load(f.read(), Loader=yaml.Loader)
+            self._add_yaml(document, os.path.basename(file_path))
+
+    def _add_yaml(self, document, file_name: Optional[str] = None) -> None:
+        if "name" not in document or "game" not in document:
+            message = "Invalid Player YAML file"
+            if file_name:
+                message = f"{message}: {file_name}"
+
+            self._show_error_in_snackbar(message)
+            return
+
+        no = len(self.players) + 1
+
+        name = document["name"]
+        game = document["game"]
```

Comment:
```suggestion
        name = document["name"]
        game = document["game"]
```
How does this handle the situation where `name` or `game` are dict and not str?

Reply:
Adding isinstance checks for this